server/app/controllers/applicant/RedirectController.java

```diff
@@ -69,37 +72,71 @@ public CompletionStage<Result> programByName(Http.Request request, String progra
       return CompletableFuture.completedFuture(result);
     }
 
-    CompletableFuture<Applicant> applicantFuture = profile.get().getApplicant();
-    CompletableFuture<Program> programFuture = programRepository.getForSlug(programName);
-
-    return CompletableFuture.allOf(applicantFuture, programFuture)
-        .thenApplyAsync(
-            empty -> {
-              if (applicantFuture.isCompletedExceptionally()) {
-                return notFound();
-              } else if (programFuture.isCompletedExceptionally()) {
-                return notFound();
-              }
-
-              Applicant applicant = applicantFuture.join();
+    return profile
+        .get()
+        .getApplicant()
+        .thenComposeAsync(
+            (Applicant applicant) -> {
               // Attempt to set default language for the applicant.
               applicant = languageUtils.maybeSetDefaultLocale(applicant);
+              final long applicantId = applicant.id;
+
               // If the applicant has not yet set their preferred language, redirect to
               // the information controller to ask for preferred language.
               if (!applicant.getApplicantData().hasPreferredLocale()) {
-                return redirect(
-                        controllers.applicant.routes.ApplicantInformationController.edit(
-                            applicant.id))
-                    .withSession(request.session().adding(REDIRECT_TO_SESSION_KEY, request.uri()));
+                return CompletableFuture.completedFuture(
+                    redirect(
+                            controllers.applicant.routes.ApplicantInformationController.edit(
+                                applicantId))
+                        .withSession(
+                            request.session().adding(REDIRECT_TO_SESSION_KEY, request.uri())));
               }
 
-              return redirect(
-                  controllers.applicant.routes.ApplicantProgramReviewController.preview(
-                      applicant.id, programFuture.join().id));
+              return getProgramVersionForApplicant(applicantId, programSlug)
```

Comment:
it feels potentially error prone to do the check in RedirectController. if we ever have another entry point the check will need to be duplicated. Is it possible to abstract it more and use the same helper here and in the normal flow?

Reply:
> if we ever have another entry point

It's hard to think of a scenario where we would -- the deep link is intended to be the canonical front door to a program for the applicant.

> Is it possible to abstract it more and use the same helper here and in the normal flow?

The tricky part here is "in the normal flow". There's a number of different routes that include the explicit program version ID, so we'd need to guard each of them. Doable, but a much larger scope of effort I don't see how an applicant would even get to one of those routes without being redirected by this controller.

I'm happy to discuss this more. If we decide we want to put deeper guards against this happening in the future I suggest we do it as a follow up though so that this fix can go out sooner.